content/en/blog/2022-06-27-unikraft-synchronization.md

```diff
@@ -0,0 +1,298 @@
+The current implementation uses the atomic operation ```__atomic_compare_exchange_n(ptr, expected, new, memorder)``` provided by GNU compiler.
+This instruction compares the *ptr* with the *expected*.
+If both are equal then it assigns the new value to the *ptr* and returns the *new* value otherwise it returns the old value of the *ptr*.
+The Unikraft library provides a wrapper named ```ukarch_compare_exchange_sync()``` on top of this atomic
+operation to simplify the use.
+
+Here is the snippet of the mutex lock and unlock function.
+The current is the pointer to the current thread structure and the owner is the owner thread of the mutex.
+The lock count is maintained to support the recursive lock.
+
+```C
+uk_mutex_lock(uk_mutex *m)
+{
+
+	/* Initialize the variable */
+
+	for (;;) {
+
+		uk_waitq_wait_event(&m->wait, m->owner == NULL);
+
+		if (ukarch_compare_exchange_sync(&m->owner, 0, current) == current) {
+			ukarch_inc(&(m->lock_count));
+			break;
+		}
+	}
+}
+
+uk_mutex_unlock(uk_mutex *m)
+{
+	if (ukarch_sub_fetch(&(m->lock_count), 1) == 0) {
+		m->owner = NULL;
+		uk_waitq_wake_up(&m->wait);
+	}
+}
+```
+
+## Semaphore
+Previously Semaphore was also disabling the interrupts for atomic operations.
+But for multiprocessor synchronization, a spinlock is added in the semaphore to protect the count variable.
+The following code snippet describes the semaphore down and up functions.
+
+```C
+uk_semaphore_down(uk_semahpore *s)
+{
+	for (;;) {
+		uk_waitq_wait_event(&s->wait, s->count > 0);
+		irqf = uk_spin_lock(&(s->sl));
+		if (s->count > 0)
+			break;
+		uk_spin_unlock_irqf(&(s->sl), irqf);
+	}
+	--s->count;
+	uk_spin_unlock(&(s->sl), irqf);
+}
+
+uk_semaphore_up(uk_semaphore *s)
+{
+
+	irqf = uk_spin_lock_irqf(&(s->sl));
+	++s->count;
+	uk_waitq_wake_up(&s->wait);
+	uk_spin_unlock_irqf(&(s->sl), irqf);
+}
+```
+
+## Reader-Writer
+The Reader writer lock allows multiple readers to simultaneously acquire the lock but only allows a single writer to acquire the lock.
+This means that the new reader can acquire the lock if there are zero writers, Whereas the writer can acquire the lock only when there is no reader as well as a writer.
+A simple implementation can be done by using the spinlock.
+But to get better performance we have implemented it using the atomic operation ```__atomic_compare_exchange_n```.
+
+Before implementing the reader-writer lock we need to consider the following problems:
+* Starvation of the writers due to readers.
+* Lost Wake-up problem.
+
+To solve the first problem, we give the higher priority to the writers i.e.  in the lock function when the reader tries to acquire the lock it first checks if there are any pending writers, if yes it sleeps in the queue.
+Also in unlock function for both reader and writer lock we first check if there are any pending writers if yes then we simply wake up the writers instead of the readers.
```

Comment:
```suggestion
Also, in `unlock()` function, both readers and writers first check if there are any pending writers.
If yes, they wake up the writers instead of the readers.
```